[V3] Add Concept Exercise: While and Switch (#1073)
exercism/javascript: concepts/conditionals-switch/about.md

# About

## General Syntax

Besides the if-statement, JavaScript also has a switch-statement to conditionally execute logic.
It is used when a single variable needs to be compared to multiple variants.
The comparison is done by checking for strict equality (`===`), see [concept comparison][concept-comparison].
For some variable `x`, the switch statement in JavaScript has the following syntax.

<!-- prettier-ignore-start -->
```javascript
switch (x) {

>   case option1:

Author: @SleeplessByte How strongly do you feel about this? Of course this is theoretically correct but I almost never see this in any real code, documentation or tutorial. So even if we explain it a bit in about.md it might be quite confusing to a student that what he/she is introduced to in the introduction.md file does not match the syntax in any other resources.

Would you also be ok if I keep the current syntax as "general" and then add a section in about.md for the block syntax? (This is also how its done on [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch#block-scope_variables_within_switch_statements).)

>   case option1:

Reviewer: I don't feel strongly about it. I have _however_ had to fix quite a few production issues because people didn't *know* about them. So not seeing it in the wild is likely more ignorance than it is intentional.

> Would you also be ok if I keep the current syntax as "general" and then add a section in about.md for the block syntax? (This is also how its done on MDN.)

Yep! I like that solution.

>   case option1:

Author: Could you give me an example of the coding bug/pitfall you saw and fixed? I am not sure I fully understand what you have in mind.

>   case option1:

Reviewer: Exactly the syntaxerror as mentioned [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/switch#block-scope_variables_within_switch_statements). 

You only see the SyntaxError if the file is loaded. In this case, the developer added a new case statement, without a block, copied the previous one, and expected it to work.

Narrator: it didn't.

>   case option1:

Author: I understand. Here is the new section: https://github.com/exercism/javascript/pull/1073/commits/7c34deddc077c5c6be5836b784731a8b296dfce6

    // code that is executed when "x === option1" is true
    break;
  case option2:
    // code that is executed when "x === option2" is true
    break;
  default:
    // code that is executed when x does not equal any of the options
}
```
<!-- prettier-ignore-end -->

The `default` case is optional and used in case you want to execute some code if none of the other options match the variable.

## Fallthrough by Default

The `break` statements above are needed because by default all cases are "fallthrough" in JavaScript.
That means without any `break` statement all the code in the cases below the first matching option would be executed even though `x` did not match those options.
This "fallthrough by default" behavior is a common pitfall when using `switch` in JavaScript.
Inside a function, `return` can also be used instead of `break` to avoid this problem.

You can use the fallthrough behavior to your advantage when you want to apply the same code for multiple cases.
You can find an example of this in the [MDN documentation][mdn-group-cases].

## Using Expressions

> Instead of a simple variable `x`, you can also put in an expression.

Reviewer: ```suggestion
Instead of a variable `x`, you can also use an expression.
```

> It will be evaluated once at the beginning of the switch statement and the result will be compared against the options.

Reviewer: ```suggestion
That expression is evaluated once at the beginning of the switch statement and the result compared against the cases.
```
